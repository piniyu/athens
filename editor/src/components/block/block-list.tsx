import styled from 'styled-components'

export const BlockListContainer = styled.div`
  display: flex;
  flex-direction: column;
`
