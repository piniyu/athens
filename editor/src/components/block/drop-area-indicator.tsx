import styled from 'styled-components'

export const DropAreaIndicatorWrap = styled.div`
  display: block;
  height: 1px;
  pointer-events: none;
  margin-bottom: -1px;
  opacity: 0.75;
  color: #0075E1;
  position: relative;
  transform-origin: left;
  z-index: 3;
  width: 100%;

  &:after {
    position: absolute;
    content: '';
    top: -0.5px;
    right: 0;
    bottom: -0.5px;
    left: calc(2em - 4px);
    border-radius: 100px;
    background: currentcolor;
  }

  &.child {
    --indent: 1.95em;
    width: calc(100% - var(--indent));
    margin-left: var(--indent)
  }
  
  &.child:after {
    border-top-left-radius: 0;
    border-bottom-left-radius: 0;
  }
  
  &.child:before {
    position: absolute;
    content: '';
    border-radius: 10em;
    border: 2px solid ;
    --size: 4px;
    width: var(--size);
    height var(--size);
    left: var(--indent);
    top: 50%;
    transform: translateY(-50%) translateX(-100%) translateX(-2px);
  }
`

export const DropAreaIndicator = ({
  child,
  style,
}: {
  child?: true
  style: React.CSSProperties
}) => {
  // const mergedStyle =
  return (
    <DropAreaIndicatorWrap
      style={style}
      className={child && 'child'}
    ></DropAreaIndicatorWrap>
  )
}
