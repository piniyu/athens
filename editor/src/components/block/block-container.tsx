import { useObservable } from '@ngneat/react-rxjs'
import React, { memo, useState } from 'react'
import { tap } from 'rxjs'
import { blockDragLeave, blockDragOver, blockDrop } from '../../handlers/drag-handlers'
import { BlockElState } from '../../interfaces'
import { blockRepo } from '../../stores/block.repository'
import { BlockContent, MemoBlockContent } from './block-content'

// const BlockRefsCountEl = (count, clickFn) => {
//   return (
//     <div>
//       <button
//         onClick={e => {
//           e.stopPropagation()
//           clickFn(e)
//         }}
//       />
//     </div>
//   )
// }

export const BlockContainer = ({ uid }: { uid: string }): JSX.Element | null => {
  const [block] = useObservable(blockRepo.getBlock$(uid)),
    [children] = useObservable(blockRepo.getBlockChildren$(uid).pipe(tap(e => console.log(`children$: ${uid}`)))),
    [state, setState] = useState<BlockElState>({
      str: {
        local: null,
        previous: null,
      },
      search: {
        type: null,
        results: [],
        query: '',
        index: -1,
      },
      dragging: false,
      dragTarget: null,
      lastKeydown: null,
      contextMenu: {
        x: null,
        y: null,
        show: false,
      },
      showEditableDom: false,
    })

  // useEffect(() => {}, [])

  if (block === undefined) {
    console.debug('bk === undefined', uid)
    return null
  }
  // const saveFn = () => {
  //   transactStateForUid(originalUid || uid, state)
  // }
  // const idleFn = () => {
  //   debounce(saveFn, 2000)
  // }
  // setState({ ...state, string: { ...state.string, saveFn, idleFn } })

  // const comp = (block, linkedRefData, opts) => {
  // const ident = [':block/uid', originalUid || uid]
  // const { uid, string, open, children, _refs } = merge(getReactiveBlockDocument(ident), block)['block']
  // const { str, open = false } = block
  // const uidSanitizedBlock = transform([
  //   // specterRecursivePath(e => e.contains('block/uid')),
  //   // () => {}
  // ])
  // const { dragging } = state
  // const isEditing = subscribe(':editing/is-editing', uid)
  // const isSelected = subscribe('::select-subs/selected?', uid)
  // const presentUser = subscribe(':presence/has-presence', uid)
  // const isPresence = seq(presentUser)

  if (state.str.previous !== block.str) {
    setState({
      ...state,
      str: { previous: block.str, local: block.str },
    })
  }

  return (
    <div
      className="block-container"
      data-uid={uid}
      data-childrenuids={children.map(e => e.uid).join(',')}
      onMouseEnter={() => {
        setState({ ...state, showEditableDom: true })
      }}
      onMouseLeave={() => {
        setState({ ...state, showEditableDom: false })
      }}
      onDragOver={e => {
        blockDragOver(e, block, state, setState)
      }}
      onDragLeave={e => {
        blockDragLeave(e, block, state, setState)
      }}
      onDrop={e => {
        blockDrop(e, block, state, setState)
      }}
    >
      {/* {state.dragTarget === 'before' && <DropAreaIndicator gridArea={'above'}/> */}

      <div className="block-body">
        {/* {children.length > 0 && (
          <Toggle
            isOpen={block.open}
            onClick={e => {
              e.stopPropagation()
              events.blockOpen(uid, !open)
            }}
          />
        )} */}
        {/* {state.contextMenu.show && <ContextMenuEl props={{ uidSanitizedBlock, state }} />} */}
        {/* <Anchor
          props={{
            isClosedWithChildren,
            block,
            shouldShowDebugDetails: reFrame10xOpen(),
            onClick = e => {
              router.navigateUid(uid, e)
            },
            onContextMenu = e => {
              bulletContextMenu(e, uid, state)
            },
            onDragStart = e => {
              bulletDragStart(e, uid, state)
            },
            onDragEnd = e => {
              bulletDragEnd(e, uid, state)
            },
          }}
        /> */}
        <BlockContent {...{ block, state, setState }} />

        {/* <InlinePresenceEl props={{ uid }} /> */}
      </div>

      {/* <InlineSearchEl props={{ block, state }} /> */}
      {/* <SlashMenuEl props={{ block, state }} /> */}

      {block.open && children.length > 0 && children.map(e => <BlockContainer key={e.uid} uid={e.uid} />)}

      {/* {state.dragTarget === ':first' && <DropAreaIndicator grid-area={'below'} child={true} />} */}
      {/* {state.dragTarget === ':after' && <DropAreaIndicator grid-area={'below'} />} */}
    </div>
  )
  // }

  // return comp
}

export const MemoBlockContainer = memo(BlockContainer)
