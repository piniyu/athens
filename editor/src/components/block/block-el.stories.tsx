import React, { useState } from 'react'
import { setEntities } from '@ngneat/elf-entities'
import { ComponentStory, ComponentMeta } from '@storybook/react'
import { blockRepo } from '../../stores/block.repository'
import { blocks } from '../../../test/helpers'
import { BlockEl } from './block-el'

/**
 * Setup data
 */
blockRepo.update([setEntities(blocks)])

export default {
  title: 'Block/BlockEl',
  component: BlockEl,
  // argTypes: {
  //   backgroundColor: { control: 'color' },
  // },
} as ComponentMeta<typeof BlockEl>

const Template: ComponentStory<typeof BlockEl> = (args) => <BlockEl {...args} />

export const Basic = Template.bind({})
Basic.args = {
  uid: 'b1',
}

// export const Editing = Template.bind({})
// Editing.args = {
//   label: 'Button',
// }

// export const Selected = Template.bind({})
// Selected.args = {
//   size: 'large',
//   label: 'Button',
// }
