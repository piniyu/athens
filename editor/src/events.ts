import { editingFocus, navigate, setCursorPosition } from './effects'
import { Block, DestructTextareaKeyEvent, Position, PositionRelation } from './interfaces'
import * as ops from './op/ops'
import { compatPosition } from './op/helpers'
import { blockRepo, blocksStore, getBlock, getBlockChildren } from './stores/block.repository'
import { rfdbRepo } from './stores/rfdb.repository'
import { genBlockUid } from './utils'
import { isInteger, isNumber } from 'lodash'

function blockSaveBlockMoveCompositeOp(sourceUid: string, refUid: string, relation: PositionRelation, str: string) {
  const location = compatPosition({ blockUid: refUid, relation })

  // TODO: composite ops
  ops.blockSaveOp(sourceUid, str)
  ops.blockMoveOp(sourceUid, location)
}

function getPrevSiblingBlockAndTargetRel(block: Block): [Block | null, PositionRelation] {
  const parent = block.parentUid ? getBlock(block.parentUid) : null,
    prevSib = parent && getNthSiblingBlock(block, parent, -1),
    targetRel = prevSib && prevSib.childrenUids.length > 0 ? 'last' : 'first'
  return [prevSib, targetRel]
}

export function up(uid: string, targetPos: number | 'end') {
  const block = getBlock(uid),
    parent = block.parentUid && getBlock(block.parentUid),
    prev = parent && getPrevBlock(block, parent),
    editingUid = prev ? prev.uid : uid
  console.debug(block, parent, prev, blocksStore.getValue())
  editingUid(editingUid, targetPos)
}

export function down(uid: string, targetPos: number | 'end') {
  const block = getBlock(uid),
    next = block && getNextBlock(block),
    editingUid = next ? next.uid : uid
  editingUid(editingUid, targetPos)
}

/**
 * "If root and 0th child, 1) if value, no-op, 2) if blank value, delete only block.
No-op if parent is missing.
No-op if parent is prev-block and block has children.
No-op if prev-sibling-block has children.
Otherwise delete block and join with previous block
If prev-block has children"
 * 
 */
export function backspace(uid: string, value: string, maybeLocalUpdates?: string) {
  // const rootEmbed = false,
  // [uid, embedId] = uidAndEmbedId(_uid),
  const block = getBlock(uid),
    { order, parentUid } = block,
    children = getBlockChildren(uid),
    parent = parentUid ? getBlock(parentUid) : null,
    prev = parent && getPrevBlock(block, parent),
    prevSib = parent && getNthSiblingBlock(block, parent, -1)

  if (
    parent === null ||
    (children.length > 0 && prevSib && prevSib.childrenUids.length > 0) ||
    (children.length > 0 && parent === prev)
  ) {
    return
  }
  if (children.length === 0 && parent.nodeTitle && order === 0 && value === '') {
    backspaceDeleteOnlyChild(uid)
    return
  }
  if (prev && maybeLocalUpdates) {
    backspaceDeleteMergeBlockWithSave(uid, value, prev, maybeLocalUpdates)
    return
  }
  if (prev) {
    backspaceDeleteMergeBlock(uid, value, prev)
  }
  console.error({ block, value, maybeLocalUpdates, children, parent, prev, prevSib })
  throw new Error('backspace::unhandled situation')
}

/**
 * "- If block is open, has children, and caret at end, create new child
- If block is CLOSED, has children, and caret at end, add a sibling block.
- If value is empty and a root block, add a sibling block.
- If caret is not at start, split block in half.
- If block has children and is closed, if at end, just add another child.
- If block has children and is closed and is in middle of block, split block.
- If value is empty, unindent.
- If caret is at start and there is a value, create new block below but keep same block index."
 */
export function enter(uid: string, dKeyDown: DestructTextareaKeyEvent) {
  const block = getBlock(uid),
    hasChildren = block.childrenUids.length > 0,
    parent = block.parentUid && getBlock(block.parentUid),
    rootBlock = (parent && parent.pageTitle) !== undefined,
    // contextRootUid = rfdbRepo.currentRoute.pathParams.id,  // contextRoot: the block opened as root
    contextRootUid = undefined,
    newUid = genBlockUid(),
    { value, start } = dKeyDown,
    caretAtEnd = value.length === start,
    valueEmpty = value.length === 0

  if (block.open && hasChildren && caretAtEnd) {
    enterAddChild(block, newUid)
  } else if (!block.open && hasChildren && caretAtEnd && parent) {
    enterNewBlock(block, parent, newUid)
  } else if ((valueEmpty || rootBlock) && parent) {
    enterNewBlock(block, parent, newUid)
  } else if (block.open && !caretAtEnd) {
    enterSplitBlock(block, newUid, value, start, 'first')
  } else if (start !== 0) {
    enterSplitBlock(block, newUid, value, start, 'after')
  } else if (valueEmpty) {
    unindent(uid, dKeyDown, '', contextRootUid)
  } else if (start === 0 && value) {
    enterBumpUp(block, newUid)
  } else {
    console.debug('[enter] ->', { uid, dKeyDown, block, parent })
    throw new Error('[enter]')
  }
}

/**
 * ;; - `block-zero`: The first block in a page
  ;; - `value`     : The current string inside the block being indented. Otherwise, if user changes block string and indents,
  ;;                 the local string  is reset to original value, since it has not been unfocused yet (which is currently the
  ;;                 transaction that updates the string).
  // if sibling block is closed with children, open
 */
export function indent(uid: string, dKeyDown: DestructTextareaKeyEvent, localStr: string) {
  const block = getBlock(uid),
    blockZero = block.order === 0,
    [sibBlock, targetRel] = getPrevSiblingBlockAndTargetRel(block)

  if (sibBlock && !blockZero) {
    const sibClosed = !sibBlock.open && sibBlock.childrenUids.length > 0,
      sibBlockOpenOp = sibClosed && blockRepo.blockOpenOp(sibBlock.uid, true),
      { start, end } = dKeyDown,
      blockSaveBlockMoveOp = blockSaveBlockMoveCompositeOp(uid, sibBlock.uid, targetRel, localStr)

    if (sibBlockOpenOp) {
      blockRepo.update([blockSaveBlockMoveOp, sibBlockOpenOp])
    } else {
      blockRepo.update([blockSaveBlockMoveOp])
    }
  }
}

export function unindent(uid: string, dKeyDown: DestructTextareaKeyEvent, localStr: string, contextRootUid?: string) {
  const block = getBlock(uid),
    parent = block.parentUid ? getBlock(block.parentUid) : null,
    doNothing = parent === null || parent.pageTitle !== null || contextRootUid === parent.uid,
    { start, end } = dKeyDown

  // async flow
  blockSaveBlockMoveCompositeOp(uid, parent.uid, 'after', localStr)
  editingUid(uid)
  setCursorPosition(uid, start, end)
}

// Backspace Events

export function backspaceDeleteMergeBlock(uid: string, value: string, prevBlock: Block) {
  // async flow
  blockRepo.blockRemoveMergeOp(uid, prevBlock.uid, value)
  // focusOnUid(prevBlock.uid, prevBlock.str.length)
  events.editingUid(uid, prevBlock.str.length)
}

export function backspaceDeleteMergeBlockWithSave(uid: string, value: string, prevBlock: Block, localUpdate?: string) {
  // async flow
  blockRepo.blockMergeWithUpdateOp(uid, prevBlock.uid, value, localUpdate)
  // focusOnUid(prevBlock.uid, localUpdate?.length)
  events.editingUid(uid, localUpdate?.length)
}

export function backspaceDeleteOnlyChild(uid: string) {
  blockRepo.blockRemoveOp(uid)
  editingUid(null)
}

// Block events

export function blockMove(sourceUid: string, targetUid: string, targetRel: PositionRelation) {
  const block = getBlock(sourceUid)
  blockRepo.update(ops.blockMoveOp(block, { blockUid: targetUid, relation: targetRel }))
}

export function blockOpen(uid: string, open: boolean) {
  ops.blockOpenOp(uid, open)
}

export function blockSave(uid: string, str: string) {
  const block = getBlock(uid),
    doNothing = block.str === str

  if (block.pageTitle) {
    throw new Error('blockSave::node-block not allow to change string')
  }
  if (!doNothing) {
    ops.blockSaveOp(uid, str)
  }
}

// Drop Events

export function dropMultiChild(sourceUids: string[], targetUid: string) {
  ops.blockMoveChain(targetUid, sourceUids, 'first')
}

/**
 * ;; When the selected blocks have same parent and are DnD under the same parent this event is fired.
    ;; This also applies if on selects multiple Zero level blocks and change the order among other Zero level blocks.
 */
export function dropMultiSiblings(sourceUids: string[], targetUid: string, dragTarget: PositionRelation) {
  ops.blockMoveChain(targetUid, sourceUids, dragTarget)
}

// Editing Events

export function editingTarget(target: HTMLTextAreaElement) {
  const uid = target.id.split('editable-uid-')[1]
  editingUid(uid)
}

export function editingUid(uid: string | null, cursorAnchor?: number | 'end') {
  rfdbRepo.updateEditingUid(uid)
  editingFocus(uid, cursorAnchor)
}

// Enter Events

export function enterAddChild(block: Block, newUid: string) {
  const position = compatPosition({ blockUid: block.uid, relation: 'first' }),
    op = ops.blockNewOp(newUid, position)
  blockRepo.update(op)
  editingUid(newUid)
}

export function enterBumpUp(block: Block, newUid: string) {
  const position = compatPosition({ blockUid: block.uid, relation: 'before' }),
    op = ops.blockNewOp(newUid, position)
  blockRepo.update(op)
  editingUid(newUid)
}

export function enterNewBlock(block: Block, parent: Block, newUid: string) {
  const op = ops.blockNewOp(newUid, { blockUid: block.uid, relation: 'after' })
  blockRepo.update(op)
  editingUid(newUid)
}

export function enterSplitBlock(
  block: Block,
  newUid: string,
  value: string,
  index: number,
  relation: PositionRelation,
) {
  const op = ops.blockSplitOp(block, newUid, value, index, relation)
  blockRepo.update(op)
  editingUid(newUid)
}

/**
 * ;; Triggered when there is a closed embeded block with no content in the top level block
  ;; and then one presses enter in the embeded block.
 */
function enterOpenBlockAddChild() {}

// Mouse Events

export function mouseDownSet() {
  rfdbRepo.updateMouseDown(true)
}

export function mouseDownUnset() {
  rfdbRepo.updateMouseDown(false)
}

// Page Events

function pageMerge() {}

export function pageNew(title: string, blockUid: string, shift?: true) {
  blockRepo.pageNewOp(title, blockUid)
  pageNewFollowup(title, shift)
  editingUid(blockUid)
}

export function pageNewFollowup(title: string, shift?: true) {
  const pageUid = blockRepo.getPageUid(title)
  if (shift) {
    // rightSidebarOpenItem(pageUid)
  } else {
    navigate({ page: { id: pageUid } })
  }
}

function pageRename(oldName: string, newName: string, callback: () => void) {
  blockRepo.renamePage(oldName, newName)
  callback()
}

// Selction Events

export function selectionSetItems(uids: string[]) {
  rfdbRepo.updateSelectionItems(uids)
}

export function selectionAddItem(uid: string, position: number | 'first' | 'last') {
  const selectedItems = rfdbRepo.getValue().selection.items,
    selectedCount = selectedItems.length

  if (position === 'last') {
    rfdbRepo.updateSelectionItems([...selectedItems, uid])
  } else if (position === 'first') {
    rfdbRepo.updateSelectionItems([uid, ...selectedItems])
  } else if (isInteger(position)) {
    if (0 <= position && position <= selectedCount) {
      const newSelectedItems = [
        ...selectedItems.slice(0, position),
        uid,
        ...selectedItems.slice(position, selectedCount),
      ]
      rfdbRepo.updateSelectionItems([...newSelectedItems, uid])
    } else {
      console.error('[selectionAddItem] Invalid insert position', position, uid, selectedItems)
    }
  }
}

export function selectionDelete() {}

export function selectionClear() {}
