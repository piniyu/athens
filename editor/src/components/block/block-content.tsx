import { useObservable } from '@ngneat/react-rxjs'
import React from 'react'
import { memo } from 'react'
import { MouseEvent } from 'react'
import { useEffect } from 'react'
import { events } from '../../events'
import { rfdbRepo } from '../../stores/rfdb.repository'
import { textareaKeyDown } from '../../textarea-keydown'

// const findSelectedItems = (e, sourceUid, targetUid) => {
//   const target = e.target,
//     page = target.closest('.node-page') || target.closes('.block-page'),
//     blocks = page.querySelectorAll('.block-container'),
//     uids = blocks.map(e => getDatasetUid(e)),
//     uidsChildrenUids = zipmap(uids, blocks.map(getDatasetChildrenUids)),
//     indexedUids = uids.mapIndexed(vector),
//     startIndex = indexedUids.filter((_idx, uid) => sourceUid === uid),
//     endIndex = indexedUids.filter((_idx, uid) => targetUid === uid),
//     selectedUids = subscribe('select-subs/items'),
//     candidateUids = indexedUids.filter(
//       (idx, _uid) => Math.min(startIndex, endIndex) <= idx <= Math.max(startIndex, endIndex),
//     ),
//     descendantUids = loop(),
//     toRemoveUids = set.intersect(selectedUids, descendantUids),
//     selectionNewUids = set.difference(candidateUids, descendantUids),
//     newSelectedUids = set.union(set.difference(selectedUids, toRemoveUids), selectionNewUids),
//     selectionOrder = indexedUids.filter((_k, v) => newSelectedUids.contains(v)).mapv('second')

//   if (startIndex && endIndex) {
//     dispatchEvent('select-events/set-items', selectionOrder)
//   }
// }

// const textareaPaste = (e, _uid, state) => {
//   const data = e.clipboardData,
//     textData = data.getData('text/plain'),
//     internalPresentation = data.getData('application/athens-representation'),
//     // internal representation
//     internal = seq(internalPresentation),
//     newUids = newUidsMap(internalPresentation),
//     reprWithNewUids = updateUids(internalPresentation, newUids),
//     // images in clipboard
//     items = arraySeq(e.clipboardData.items),
//     { head, tail } = destructTarget(e.target),
//     imgRegex = /#"(?i)^image\/(p?jpeg|gif|png)$"/,
//     callback = () => {},
//     // external to internal representation
//     textToInter = textData !== '' && textToInternalPresentation(textData),
//     lineBreaks = reFind(/\r?\n/, textData),
//     noShift = state.lastKeydown !== 'shift'

//   if (internal) {
//     e.preventDefaul()
//     dispatchEvent('paste-internal', uid, state.string.local, reprWithNewUids)
//   } else if (seq(filter())) {
//     // For images
//   } else if (lineBreaks && noShift) {
//     e.preventDefaul()
//     dispatchEvent('paste-internal', uid, state.string.local, textToInter)
//   } else if (noShift) {
//     e.preventDefaul()
//     dispatchEvent('paste-verbatim', uid, textData)
//   }
// }

function textareaChange(
  event: React.ChangeEvent<HTMLTextAreaElement>,
  _uid: string,
  state: BlockElState,
  setState: BlockElStateSetFn,
) {
  setState({
    ...state,
    str: { ...state.str, local: event.target.value },
  })
  // if (state.string.idleFn) {
  //   state.string.idleFn()
  // }
}

function textareaClick(event: MouseEvent, targetUid: string): void {
  // const [targetUid] = uidAndEmbedId(targetUid),
  // const sourceUid = subscribe('editing/uid'),
  //   shift = e.shiftKey
  // if (shift && sourceUid && targetUid && sourceUid !== targetUid) {
  //   findSelectedItems(e, sourceUid, targetUid)
  //   dispatchEvent('select-events/clear')
  // }
  const shift = event.shiftKey
  rfdbRepo.editingUid$.subscribe(sourceUid => {
    if (shift && sourceUid && targetUid && sourceUid !== targetUid) {
      // findSelectedItems(e, sourceUid, targetUid)
      // dispatchEvent('select-events/clear')
    }
  })
}

function globalMouseup() {
  document.removeEventListener('mouseup', globalMouseup)
  rfdbRepo.setProps({ mouseDown: false })
}

function textareaMouseDown(e: MouseEvent) {
  e.stopPropagation()
  if (!e.shiftKey) {
    events.editingTarget(e.target as HTMLTextAreaElement)

    const { mouseDown } = rfdbRepo.getValue()
    if (!mouseDown) {
      rfdbRepo.setProps({ mouseDown: true })
      document.addEventListener('mouseup', globalMouseup)
    }
  }
}

function textareaMouseEnter() {
  const { editing, mouseDown } = rfdbRepo.getValue(),
    sourceUid = editing?.uid

  if (mouseDown) {
    // dispatchEvent('select-events/clear')
    // findSelectedItems(e, sourceUid, targetUid)
    rfdbRepo.setProps({ selection: { items: [] } })
  }
}

// View

export const BlockContent = ({
  block,
  state,
  setState,
}: {
  block: Block
  state: BlockElState
  setState: React.Dispatch<React.SetStateAction<BlockElState>>
}) => {
  // const { uid, originalUid, header } = block.block,
  // selectedItems = subscribe('select-subs/items')

  const { uid } = block,
    [editing] = useObservable(rfdbRepo.getBlockIsEditing$(uid))

  useEffect(() => {
    // console.debug('BlockContentEl::enter ' + uid)
  })

  return (
    <div className="block-content">
      {(state.showEditableDom || editing) && (
        <textarea
          value={state.str.local ?? ''}
          // className={['textarea', empty(selectedItems) && editing ? 'is-editing' : undefined].join('')}
          id={`editable-uid-${uid}`}
          onChange={e => {
            textareaChange(e, uid, state, setState)
          }}
          // onPaste={e => {
          //   textareaPaste(e, uid, state)
          // }}
          onKeyDown={e => {
            textareaKeyDown(e, uid, editing, state, setState)
          }}
          // onBlur={state.string.saveFn}
          onClick={e => {
            textareaClick(e, uid)
          }}
          onMouseEnter={e => {
            textareaMouseEnter()
          }}
          onMouseDown={e => {
            textareaMouseDown(e)
          }}
        />
      )}

      {/* {parseAndRender(state.string.local, originalUid || uid)} */}
      <span>
        {state.str.local}xxxxxxxxxxxx{uid}
      </span>
    </div>
  )
}

export const MemoBlockContent = memo(BlockContent)
