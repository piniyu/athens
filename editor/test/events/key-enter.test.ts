import { addEntities } from '@ngneat/elf-entities'
import { diff } from 'deep-object-diff'
import { events } from '../../src/events'
import { blockRepo, blocksStore, getBlock } from '../../src/stores/block.repository'
import { rfdbRepo } from '../../src/stores/rfdb.repository'

function clean(cur: Record<string, Block>) {
  for (const [k, v] of Object.entries(cur)) {
    delete cur[k].editTime
  }
  return cur
}

/**
 *  0
 *  - 1
 *    - 3
 *  - 2
 */

const blocks: Block[] = [
  { uid: '0', str: '0', order: 0, parentUid: null, childrenUids: ['1', '2'] },
  { uid: '1', str: '1', order: 0, parentUid: '0', childrenUids: ['3'] },
  { uid: '2', str: '2', order: 1, parentUid: '0', childrenUids: [] },
  { uid: '3', str: '3', order: 0, parentUid: '1', childrenUids: [] },
]

const [p, b1, b2, b3] = blocks

beforeEach(() => {
  blockRepo.update([addEntities(blocks)])
})

it('enterAddChild()', () => {
  rfdbRepo.setProps({ editing: { uid: '1' } })

  /**
   *  0
   *  (- x)
   *  - 1
   *    - 3
   *  - 2
   */
  let cur = blocksStore.getValue().entities,
    next

  events.enterAddChild(p, 'x')
  next = clean(blocksStore.getValue().entities)
  expect(diff(cur, next)).toMatchInlineSnapshot(`
    Object {
      "0": Object {
        "childrenUids": Object {
          "0": "x",
          "1": "1",
          "2": "2",
        },
      },
      "1": Object {
        "order": 1,
      },
      "2": Object {
        "order": 2,
      },
      "x": Object {
        "childrenUids": Array [],
        "open": true,
        "order": 0,
        "parentUid": "0",
        "str": "",
        "uid": "x",
      },
    }
  `)

  /**
   *  0
   *  - x
   *    (- y)
   *  - 1
   *    - 3
   *  - 2
   */
  const x = getBlock('x')
  events.enterAddChild(x, 'y')
  cur = next
  next = clean(blocksStore.getValue().entities)
  expect(diff(cur, next)).toMatchInlineSnapshot(`
    Object {
      "x": Object {
        "childrenUids": Object {
          "0": "y",
        },
      },
      "y": Object {
        "childrenUids": Array [],
        "open": true,
        "order": 0,
        "parentUid": "x",
        "str": "",
        "uid": "y",
      },
    }
  `)

  /**
   *  0
   *  - x
   *    - y
   *      (- z)
   *  - 1
   *    - 3
   *  - 2
   */
  const y = getBlock('y')
  events.enterAddChild(y, 'z')
  cur = next
  next = clean(blocksStore.getValue().entities)
  expect(diff(cur, next)).toMatchInlineSnapshot(`
    Object {
      "y": Object {
        "childrenUids": Object {
          "0": "z",
        },
      },
      "z": Object {
        "childrenUids": Array [],
        "open": true,
        "order": 0,
        "parentUid": "y",
        "str": "",
        "uid": "z",
      },
    }
  `)

  /**
   *  0
   *  - x
   *    (- y2)
   *    - y
   *      (- z)
   *  - 1
   *    - 3
   *  - 2
   */
  events.enterAddChild(x, 'y2')
  cur = next
  next = clean(blocksStore.getValue().entities)
  expect(diff(cur, next)).toMatchInlineSnapshot(`
    Object {
      "x": Object {
        "childrenUids": Object {
          "0": "y2",
          "1": "y",
        },
      },
      "y": Object {
        "order": 1,
      },
      "y2": Object {
        "childrenUids": Array [],
        "open": true,
        "order": 0,
        "parentUid": "x",
        "str": "",
        "uid": "y2",
      },
    }
  `)
})

it('enterNewBlock()', () => {
  rfdbRepo.setProps({ editing: { uid: '0' } })

  const parent = blocks[0],
    block = blocks[1]

  // events.up('1', 'end')
  events.enterNewBlock(block, parent, 'x')
  expect(blocksStore.getValue().entities).toMatchSnapshot()

  // expect(rfdbRepo.getValue().editing?.uid).toEqual('1')
})

it('enterNewBlock, enterNewBlock, up', () => {
  rfdbRepo.setProps({ editing: { uid: '0' } })

  const parent = blocks[0]

  // events.up('1', 'end')
  events.enterNewBlock(blocks[1], parent, 'x')
  expect(rfdbRepo.getValue().editing?.uid).toEqual('x')

  events.up('x', 0)
  expect(rfdbRepo.getValue().editing?.uid).toEqual('1')

  events.enterNewBlock(getBlock('x'), parent, 'y')
  expect(blocksStore.getValue().entities).toMatchSnapshot()
  expect(rfdbRepo.getValue().editing?.uid).toEqual('y')

  events.up('y', 0)
  expect(rfdbRepo.getValue().editing?.uid).toEqual('x')

  // expect(rfdbRepo.getValue().editing?.uid).toEqual('1')
})
