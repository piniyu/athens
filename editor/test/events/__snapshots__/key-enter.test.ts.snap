// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`enterNewBlock() 1`] = `
Object {
  "0": Object {
    "childrenUids": Array [
      "1",
      "x",
      "2",
    ],
    "editTime": 1647622347516,
    "order": 0,
    "parentUid": null,
    "str": "0",
    "uid": "0",
  },
  "1": Object {
    "childrenUids": Array [
      "3",
    ],
    "order": 0,
    "parentUid": "0",
    "str": "1",
    "uid": "1",
  },
  "2": Object {
    "childrenUids": Array [],
    "order": 2,
    "parentUid": "0",
    "str": "2",
    "uid": "2",
  },
  "3": Object {
    "childrenUids": Array [],
    "order": 0,
    "parentUid": "1",
    "str": "3",
    "uid": "3",
  },
  "x": Object {
    "childrenUids": Array [],
    "editTime": 1647622347516,
    "open": true,
    "order": 1,
    "parentUid": "0",
    "str": "",
    "uid": "x",
  },
  "y": Object {
    "childrenUids": Array [
      "z",
    ],
    "open": true,
    "order": 1,
    "parentUid": "x",
    "str": "",
    "uid": "y",
  },
  "y2": Object {
    "childrenUids": Array [],
    "open": true,
    "order": 0,
    "parentUid": "x",
    "str": "",
    "uid": "y2",
  },
  "z": Object {
    "childrenUids": Array [],
    "open": true,
    "order": 0,
    "parentUid": "y",
    "str": "",
    "uid": "z",
  },
}
`;

exports[`enterNewBlock, enterNewBlock, up 1`] = `
Object {
  "0": Object {
    "childrenUids": Array [
      "1",
      "x",
      "y",
      "2",
    ],
    "editTime": 1647622347519,
    "order": 0,
    "parentUid": null,
    "str": "0",
    "uid": "0",
  },
  "1": Object {
    "childrenUids": Array [
      "3",
    ],
    "order": 0,
    "parentUid": "0",
    "str": "1",
    "uid": "1",
  },
  "2": Object {
    "childrenUids": Array [],
    "order": 3,
    "parentUid": "0",
    "str": "2",
    "uid": "2",
  },
  "3": Object {
    "childrenUids": Array [],
    "order": 0,
    "parentUid": "1",
    "str": "3",
    "uid": "3",
  },
  "x": Object {
    "childrenUids": Array [],
    "editTime": 1647622347517,
    "open": true,
    "order": 1,
    "parentUid": "0",
    "str": "",
    "uid": "x",
  },
  "y": Object {
    "childrenUids": Array [],
    "editTime": 1647622347519,
    "open": true,
    "order": 2,
    "parentUid": "0",
    "str": "",
    "uid": "y",
  },
  "y2": Object {
    "childrenUids": Array [],
    "open": true,
    "order": 0,
    "parentUid": "x",
    "str": "",
    "uid": "y2",
  },
  "z": Object {
    "childrenUids": Array [],
    "open": true,
    "order": 0,
    "parentUid": "y",
    "str": "",
    "uid": "z",
  },
}
`;
